### PR TITLE
feat: Mask sequences with high logprob error

### LINKED ISSUE #1910: [super-pr] masking by seq level logprob error
ffcb0d176f6ecc5a6761ffb287fdc3692cb78fd4

possibly dup of #1838

### PR DESCRIPTION
**Pull request by yfw**
# What does this PR do ?

Adds `seq_logprob_error_threshold` which can be used to mask sequences that exceed the specified sequence level logprob error. This was used for nano-v3 training.

# Issues
List issues that this PR closes ([syntax](https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)):
closes #1910

# Usage
* **You can potentially add a usage example below**

```python
# Add a code snippet demonstrating how to use this
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](/NVIDIA-NeMo/RL/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA-NeMo/RL/blob/main/docs/testing.md) for how to run tests
- [ ] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA-NeMo/RL/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information
* ...

## Summary by CodeRabbit

* **New Features**
  * Added configurable sequence-level error masking for GRPO training
  * New metrics to monitor masked sequences and error statistics

* **Tests**
  * Added comprehensive test suite for sequence error masking functionality

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: tests/unit/algorithms/test_grpo.py [hunk 1]
```diff
@@ -1898,3 +1899,317 @@ def test_validate_returns_empty_when_no_dataloader(self):
...
+            prev_logprobs,
+            generation_logprobs,
+            token_mask=token_mask,
+        )
+        rewards = torch.tensor([1.0, 1.0, 1.0])
+
+        # Threshold of 1.5 should mask sequence 2 (exp(0.41) > 1.5)
+        max_error, num_masked, masked_pct = compute_and_apply_seq_logprob_error_masking(
+            train_data, rewards, seq_logprob_error_threshold=1.5
+        )
```
[the remark this answers is not in the document]
Comment (yfw) [reply tests/unit/algorithms/test_grpo.py]: Addressed in d07df80b9e649adee754021172e5ad44c1ca8d17

File: nemo_rl/algorithms/grpo.py [hunk 2]
```diff
@@ -1161,6 +1163,89 @@ def _log_mixed_rewards_and_advantages_information(
...
+
+    Returns:
+        Tuple of (max_seq_mult_prob_error, num_masked_seqs, masked_correct_pct)
+    """
+    # Compute sequence-level logprob error metrics (always)
+    token_mask = train_data["token_mask"][:, 1:]
+    sample_mask = train_data["sample_mask"]
+    prev_logprobs = train_data["prev_logprobs"][:, 1:]
+    generation_logprobs = train_data["generation_logprobs"][:, 1:]
+    lp_error = torch.abs(generation_logprobs - prev_logprobs)
```
Comment (terrykong) [inline nemo_rl/algorithms/grpo.py:1194]: i thought MIS was an upper bound masking of train/gen? this looks two sided b/c of abs value.

should we expose two knobs so we can always compare the framework with another FW that has textbook MIS vs our variant?

**Review by terrykong — COMMENTED** [review]
small comment, but generally lgtm.

could you also add this to the docs since we now support MIS after this?

File: nemo_rl/algorithms/grpo.py (hunk 2 above)
Comment (yfw) [reply nemo_rl/algorithms/grpo.py:1194]: This is different than MIS because we take the _absolute value_ of the difference between `prev_logprobs` and `generation_logprobs`, whereas MIS just uses the difference between `prev_logprobs` and `generation_logprobs`. Because we then use `torch.exp` to transform back to probability space, setting lower and upper bounds on the difference would not let us recover MIS. I added a comment about the difference between this and MIS in the config definition.
